circuits/circuits/claiming/distribute.circom

```diff
@@ -0,0 +1,229 @@
+pragma circom 2.1.5;
+
+include "../node_modules/circomlib/circuits/poseidon.circom"; // TODO: consider Poseidon2
+include "../node_modules/circomlib/circuits/gates.circom";
+include "../node_modules/circomlib/circuits/comparators.circom";
+include "../merkle_tree.circom";
+include "./utils.circom";
+include "./challenge.circom";
+
+// Distrubte takes a commitment to all claims, validates them, and outputs a commitment to the rightful owners
+// 
+//      1/  First, we create the "filtered" tree. Every valid claim is added to the filtered tree,
+//          unless the prover provides a challenge claim that supersedes it.
+//      2/  Then we prove that the "sorted" tree contains every value from the filtered tree and vice versa.
+//          This proves that they are permutations of each other, except if there are duplicates, which may differ between trees.
+//      3/  Then we prove that every coin range in the sorted tree is strictly greater than the last
+//          I.e., for ranges [a,b] and [c,d], that b < c.
+//          This proves that there are no duplicates
+//
+// We know that all claims in the sorted tree are disjoint because of step 3, and we know that every valid unchallengable claim is in the sorted tree, because
+// they can only be removed by being invalid or challenged. So we know that the sorted tree consists of the actual exactly the rightful owners (according to the claims).
```

Comment:
This is not actually true! I was assuming that if a claim was challenged, then the challenger would take its place in the tree, but actually the challenger might also be invalid leaving *no one else* to own the coins due to overlaps.

Suppose there are the following valid claims:
 1/ `A owned [1,10] at block 1`
 2/ `B owned [6,15] at block 2`
 3/ `C owned [11,20] at block 3`

Notice 3 supersedes 2, and 2 supersedes 1 but 3 doesn't supersede 1. This means that we can have a filtered list containing just claim 3, or claims 3 and 1.

The proper owners are: `A owns [1,5]`, `B owns [6,10]`, `C owns [11, 20]`. So we need to do some kind of splitting.

Reply:
i think a good approach forward for this PR is to first write a spec/draw diagrams to outline the algorithm. This is a crucial one to get right and also a big one. We can then split the review by first discussing that and then the implementation.